### ISSUE TITLE
Different Hashes for Same Image

### ISSUE THREAD

**Issue by JohnTravolski** [JohannesBuchner/imagehash#89]
I find it concerning that the results of the hashes used for images can depend on the library used to read in the file.  I have only had this issue with .jpg images, so I suspect these two libraries don't decode them the same way.

```
from PIL import Image
import numpy as np
import cv2
import imagehash

# In this code, the same image is read in with two different libraries.  
# The perceptual hash is then calculated for both, and they differ.

# convert the image read in from PIL to a cv2-style numpy array
pic = Image.open("original.jpg")
pil_to_cv2 = np.array(pic)
pil_to_cv2 = cv2.cvtColor(pil_to_cv2, cv2.COLOR_RGB2BGR)
cv2.imwrite("pil_to_cv2.png", pil_to_cv2)

# convert the image read in from cv2 to a PIL-style image
cv2img = cv2.imread("original.jpg", cv2.IMREAD_UNCHANGED)
cv2_to_pil_arr = cv2.cvtColor(cv2img, cv2.COLOR_BGR2RGB)
cv2_to_pil = Image.fromarray(cv2_to_pil_arr)
cv2_to_pil.save("cv2_to_pil.png")

phash1 = str(imagehash.phash(cv2_to_pil))
phash2 = str(imagehash.phash(pic))

print(phash1)
print(phash2)
```
Output:
```
dbae2805665574f1
cbfa2c04375b74b0
```

Apparently there are different ways that certain libraries decode JPEG images; some take "shortcuts" which actually decode to a different result, resulting in different perceptual hashes of the data.  This code shows that, whenever I load in an image with CV2 and perceptual hash it, I get a different hash than when I load in the image with PIL and hash it.

One interesting point is that the perceptual hashes seem to differ between the two on images that have HIGH amounts of symmetry, such as images with lots of vertical stripes or images that have been mirrored across both axes.  This usually results in two very different hashes.  For images with no symmetry, the phashes usually don't differ at all, and on the rare occasion that they do, it's only by a few bits.

This is the offending image:

![original](https://user-images.githubusercontent.com/36255053/59570495-6954c700-905e-11e9-9c50-8d26bb0ba904.jpg)

In case that upload has been recompressed, potentially altering the results, I have uploaded the file here:

https://drive.google.com/file/d/1Gy4-uPXIWDKsWkgogi2fuhX0B11N3OUj/view?usp=sharing

**Comment by JohannesBuchner** [comment]
Interesting!

**Comment by btxgit** [comment]
Wow, I've hashed over 15TB of JPEG images with both  libraries, and I have yet to see any major decode differences like you experienced above.  I've been working with some golang code code that usually produces the identical hash, but sometimes there's single bit differences in the identical image.

I went through the various sources of difference, eventually found it was due to the grayscale conversion algorithm.  Both PIL and the golang package use the standard rgb -> grayscale alg:

(r*299 + g*587 + b*114)/1000

But wait, PIL is using // rather than /.  That causes a truncation of the result, resulting in a rounding error.  But honestly, I don't know the source of this algorithm, so I don't know 

The "fix", if you choose to call it that, is:
in ImageColor.py, change:
color = (r*299 + g*587 + b*114)//1000
to
color = round((r*299 + g*587 + b*114)/1000)

Since I already had several hundred TB hashed with the alt dhash alg, I modified the other algorithm to produce the same results  I posted this on the Pillow github as well.

Anyway, thank you so much for this software!
